### PR TITLE
db: migration script for key format change

### PR DESCRIPTION
**Pull request by tychoish**
This just includes the migration code and pretty extensive
tests. We'll have to integrate this into some CLI tool, or potentially
into the SDK rather than tendermint propper. I think it makes sense to
deal with the integration/operationalization questions seperate from
the migration itself.

### CODE REVIEW

**Comment by marbar3778** [comment]
How will this script be run? It seems to be we would need to have a state migrate cmd or onStart check some state version and run it ourselves. 

How long do you see this running on a db of a couple gbs?

**Comment by tychoish** [comment]
> How long do you see this running on a db of a couple gbs?

No clue, we should definitely test that more. I think the speed is more dependent on the number of keys rather than the total data size, though clearly those two are related.

The script forces an fsync (or similar) 10% of the time, and we could definitely do that less frequently, which speed things up, but it's a trade off, of course. 

> Will we revert other breaking changes for 0.35?

Likely? Unless we can mitigate or migrate the break?

File: scripts/keymigrate/migrate.go [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+		}
+
+		return orderedcode.Append(nil, int64(0), int64(val))
+	case bytes.HasPrefix(key, keyID("P:")):
+		val, err := strconv.Atoi(string(key[2:]))
+		if err != nil {
+			return nil, err
+		}
+
+		return orderedcode.Append(nil, int64(1), int64(val))
```
Comment (cmwaters) [inline scripts/keymigrate/migrate.go:115]: Block part has both a height and an index:
```go
// v0.34
func calcBlockPartKey(height int64, partIndex int) []byte {
	return []byte(fmt.Sprintf("P:%v:%v", height, partIndex))
}
```
```go
// v0.35
func blockPartKey(height int64, partIndex int) []byte {
	key, err := orderedcode.Append(nil, prefixBlockPart, height, int64(partIndex))
	if err != nil {
		panic(err)
	}
	return key
}
```

File: scripts/keymigrate/migrate.go [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+			return nil, err
+		}
+
+		return orderedcode.Append(nil, int64(7), int64(val))
+	case bytes.HasPrefix(key, keyID("stateKey")):
+		return orderedcode.Append(nil, int64(8))
+	case bytes.HasPrefix(key, []byte{0x00}): // committed evidence
+		return orderedcode.Append(nil, int64(9))
+	case bytes.HasPrefix(key, []byte{0x01}): // pending evidence
+		return orderedcode.Append(nil, int64(10))
```
Comment (cmwaters) [inline scripts/keymigrate/migrate.go:163]: Will this work? Won't you will need to extract out the height and hash
```go
// v0.34
func keyPending(evidence types.Evidence) []byte {
	return append([]byte{baseKeyPending}, keySuffix(evidence)...)
}

func keySuffix(evidence types.Evidence) []byte {
	return []byte(fmt.Sprintf("%s/%X", bE(evidence.Height()), evidence.Hash()))
}
```
```go
// v0.35
func keyPending(evidence types.Evidence) []byte {
	var height int64 = evidence.Height()
	key, err := orderedcode.Append(nil, prefixPending, height, string(evidence.Hash()))
	if err != nil {
		panic(err)
	}
	return key
}
```

**Review by cmwaters — COMMENTED** [review]
Nice preliminary work!

I agree that we will need to add this to the CLI. I am actually not too sure how Cosmos SDK will handle this migration (I think either they will import the command or the `Migrate` function)

**Comment by tychoish** [comment]
I think I got all of @cmwaters's (spectacular!) catches, including user/application tx.index keys that might have `/`s in the string component.

File: scripts/keymigrate/migrate.go [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+		keyID("SC:"),
+		keyID("BH:"),
+		// light
+		keyID("size"),
+		keyID("lb/"),
+		// evidence
+		keyID([]byte{0x00}),
+		keyID([]byte{0x01}),
+		// tx index
+		keyID("tx.height/"),
```
Comment (cmwaters) [inline scripts/keymigrate/migrate.go:94]:
```suggestion
		keyID("tx.height/"),
		keyID("tx.hash/"),
```

File: scripts/keymigrate/migrate_test.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+		"Height":            makeKey(t, int64(0), int64(val)),
+		"BlockPart":         makeKey(t, int64(1), int64(val), int64(val)),
+		"BlockPartTwo":      makeKey(t, int64(1), int64(val+2), int64(val+val)),
+		"BlockCommit":       makeKey(t, int64(2), int64(val)),
+		"SeenCommit":        makeKey(t, int64(3), int64(val)),
+		"BlockHeight":       makeKey(t, int64(4), int64(val)),
+		"Validators":        makeKey(t, int64(5), int64(val)),
+		"ConsensusParams":   makeKey(t, int64(6), int64(val)),
+		"ABCIResponse":      makeKey(t, int64(7), int64(val)),
+		"State":             makeKey(t, int64(8), int64(val)),
```
Comment (cmwaters) [inline scripts/keymigrate/migrate_test.go:55]:
```suggestion
		"State":             makeKey(t, int64(8)),
```

File: scripts/keymigrate/migrate_test.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+		// we want to make sure that the key space for new and
+		// legacy keys are entirely non-overlapping.
+
+		legacyPrefixes := getLegacyPrefixKeys(42)
+
+		newPrefixes := getNewPrefixKeys(t, 42)
+
+		require.Equal(t, len(legacyPrefixes), len(newPrefixes))
+
+		t.Run("LegacyEvidence", func(t *testing.T) {
```
Comment (cmwaters) [inline scripts/keymigrate/migrate_test.go:101]: Why do you have the Evidence suffix here? This is testing all keys right? not just evidence ones

**Review by cmwaters — COMMENTED** [review]
I think this almost ready. Would be eager to try this out on an actual node

**Review by cmwaters — APPROVED** [review]
I think this is ready to be migrated into master. 😄 

One minor question: How long do we foresee this taking? Does it make sense to show some sort of status so users know that it is working?

**Pull request merged**